crates/wasi-common/yanix/src/file.rs

```diff
@@ -0,0 +1,191 @@
+        libc::openat(
+            dirfd,
+            path.as_ptr(),
+            oflag.bits(),
+            libc::c_uint::from(mode.bits()),
+        )
+    })
+}
+
+pub fn readlinkat<P: AsRef<OsStr>>(dirfd: RawFd, path: P) -> Result<OsString> {
+    let path = CString::new(path.as_ref().as_bytes())?;
+    let buffer = &mut [0u8; libc::PATH_MAX as usize + 1];
+    Errno::from_result(unsafe {
+        libc::readlinkat(
+            dirfd,
+            path.as_ptr(),
+            buffer.as_mut_ptr() as *mut _,
+            buffer.len(),
+        )
+    })
+    .and_then(|nread| {
+        let link = OsStr::from_bytes(&buffer[0..nread.try_into()?]);
+        Ok(link.into())
+    })
+}
+
+pub fn mkdirat<P: AsRef<OsStr>>(dirfd: RawFd, path: P, mode: Mode) -> Result<()> {
+    let path = CString::new(path.as_ref().as_bytes())?;
+    Errno::from_success_code(unsafe { libc::mkdirat(dirfd, path.as_ptr(), mode.bits()) })
+}
+
+pub fn linkat<P: AsRef<OsStr>>(
+    old_dirfd: RawFd,
+    old_path: P,
+    new_dirfd: RawFd,
+    new_path: P,
+    flags: AtFlag,
+) -> Result<()> {
+    let old_path = CString::new(old_path.as_ref().as_bytes())?;
+    let new_path = CString::new(new_path.as_ref().as_bytes())?;
+    Errno::from_success_code(unsafe {
+        libc::linkat(
+            old_dirfd,
+            old_path.as_ptr(),
+            new_dirfd,
+            new_path.as_ptr(),
+            flags.bits(),
+        )
+    })
+}
+
+pub fn unlinkat<P: AsRef<OsStr>>(dirfd: RawFd, path: P, flags: AtFlag) -> Result<()> {
+    let path = CString::new(path.as_ref().as_bytes())?;
+    Errno::from_success_code(unsafe { libc::unlinkat(dirfd, path.as_ptr(), flags.bits()) })
+}
+
+pub fn renameat<P: AsRef<OsStr>>(
+    old_dirfd: RawFd,
+    old_path: P,
+    new_dirfd: RawFd,
+    new_path: P,
+) -> Result<()> {
+    let old_path = CString::new(old_path.as_ref().as_bytes())?;
+    let new_path = CString::new(new_path.as_ref().as_bytes())?;
+    Errno::from_success_code(unsafe {
+        libc::renameat(old_dirfd, old_path.as_ptr(), new_dirfd, new_path.as_ptr())
+    })
+}
+
+pub fn symlinkat<P: AsRef<OsStr>>(old_path: P, new_dirfd: RawFd, new_path: P) -> Result<()> {
+    let old_path = CString::new(old_path.as_ref().as_bytes())?;
+    let new_path = CString::new(new_path.as_ref().as_bytes())?;
+    Errno::from_success_code(unsafe {
+        libc::symlinkat(old_path.as_ptr(), new_dirfd, new_path.as_ptr())
+    })
+}
+
+pub fn fstatat<P: AsRef<OsStr>>(dirfd: RawFd, path: P, flags: AtFlag) -> Result<libc::stat> {
+    use std::mem::MaybeUninit;
+    let path = CString::new(path.as_ref().as_bytes())?;
+    let mut filestat = MaybeUninit::<libc::stat>::uninit();
+    Errno::from_result(unsafe {
+        libc::fstatat(dirfd, path.as_ptr(), filestat.as_mut_ptr(), flags.bits())
+    })?;
+    Ok(unsafe { filestat.assume_init() })
+}
+
+/// `fionread()` function, equivalent to `ioctl(fd, FIONREAD, *bytes)`.
+pub fn fionread(fd: RawFd) -> Result<usize> {
```

Comment:
This one too, and a few others as well :-).

Reply:
Yep, lemme adjust all remaining calls ;-)

Reply:
Done! Hopefully, this time I got them all!